influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -409,61 +526,40 @@ fn background_cache_request_handler(
 ) -> tokio::task::JoinHandle<()> {
     tokio::spawn(async move {
         while let Some(CacheRequest { path, notifier }) = rx.recv().await {
-            // clone the path before acquiring the lock:
-            let path_cloned = path.clone();
-            // Check that the cache does not already contain an entry for the provide path, or that
-            // it is not already in the process of fetching the given path:
-            let mut cache_lock = mem_store.cache.lock().await;
-            if cache_lock
-                .get(&path)
-                .is_some_and(|entry| entry.is_fetching() || entry.is_success())
-            {
+            // We assume that objects on object store are immutable, so we can skip objects that
+            // we have already fetched:
+            if mem_store.cache.path_already_fetched(&path).await {
                 continue;
             }
             // Put a `Fetching` state in the entry to prevent concurrent requests to the same path:
-            let _ = cache_lock.put_with_weight(path_cloned, CacheEntry::Fetching);
-            // Drop the lock before spawning the task below
-            drop(cache_lock);
+            mem_store.cache.set_fetching(&path).await;
             let mem_store_captured = Arc::clone(&mem_store);
             tokio::spawn(async move {
-                let cache_insertion_result = match mem_store_captured.inner.get(&path).await {
+                match mem_store_captured.inner.get(&path).await {
                     Ok(result) => {
                         let meta = result.meta.clone();
                         match result.bytes().await {
-                            Ok(data) => mem_store_captured.cache.lock().await.put_with_weight(
-                                path,
-                                CacheEntry::Success(Arc::new(CacheValue { data, meta })),
-                            ),
-                            Err(error) => {
-                                error!(%error, "failed to retrieve payload from object store get result");
-                                mem_store_captured
+                            Ok(data) => {
+                                if let Err(error) = mem_store_captured
                                     .cache
-                                    .lock()
+                                    .set_success(&path, CacheValue { data, meta })
                                     .await
-                                    .put_with_weight(path, CacheEntry::Failed)
+                                {
+                                    error!(%error, "failed to set the success state on the cache");
+                                    mem_store_captured.cache.remove(&path).await;
```

Comment:
Might be better to do the remove in `set_success` directly, so that it does it all under one lock.